### ISSUE TITLE
Converting multiple std::vector<vec_t> to std::vector<tensor_t>

### ISSUE THREAD

**Issue by shivak7** [tiny-dnn/tiny-dnn#684]
I am trying to modify my sequential NN network to a graph network that takes in two input images and converges at the first fully connected layer. I was unable to find any clear examples for loading multiple inputs. Is there a way to convert multiple vec_t vectors to a single tensor_t vector?

**Comment by shivak7** [comment]
So I manged to push 2 vec_t elements into a tensor_t per training sample. That seems to work fine. But then when I try to do training:
`net.train<tiny_dnn::cross_entropy>(opt, Data, Labels, 20, 30, []() {}, [&]()...`

 Here Data is a tensor_t object and Labels is a label_t object, I get compilation errors saying there are no matching function calls:

`candidate: template<class Error, class Optimizer, class OnBatchEnumerate, class OnEpochEnumerate> bool tiny_dnn::network<NetType>::train(Optimizer&, const std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > >&, const std::vector<unsigned int>&, size_t, int, OnBatchEnumerate, OnEpochEnumerate, bool, int, const std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > >&) [with Error = Error; Optimizer = Optimizer; OnBatchEnumerate = OnBatchEnumerate; OnEpochEnumerate = OnEpochEnumerate; NetType = tiny_dnn::graph]`

**Comment by shivak7** [comment]
So I went through network.h and realized the only option I ever had was to use fit instead of train and it compiles. But I wonder if this is ok because I am only using single value class labels.

Also, the test function now throws similar errors as above. How would one test with std::vector<tensor_t> type inputs?

**Comment by shivak7** [comment]
I tried training in the mean time using the graph structure shown here. But I get a segfault. Backtracing gives me this. Does the fit function crash if one uses labels instead of vectors?

 0x0000000000419c89 in tiny_dnn::layer::out_data_size (this=0x7fffffffadb0) at ./tiny_dnn/layers/layer.h:133
#2  0x00000000004ad22b in tiny_dnn::nodes::out_data_size (this=0x7fffffffdd00) at ./tiny_dnn/nodes.h:130
#3  0x00000000004ad252 in tiny_dnn::nodes::label2vec (this=0x7fffffffdd00, t=0x86af70, num=512, vec=std::vector of length 0, capacity 0)
    at ./tiny_dnn/nodes.h:181
#4  0x00000000004ad398 in tiny_dnn::nodes::label2vec (this=0x7fffffffdd00, labels=std::vector of length 512, capacity 512 = {...}, 
    vec=std::vector of length 0, capacity 0) at ./tiny_dnn/nodes.h:194
#5  0x00000000004b8348 in tiny_dnn::network<tiny_dnn::graph>::normalize_tensor (this=0x7fffffffdce0, 
    inputs=std::vector of length 512, capacity 512 = {...}, normalized=std::vector of length 0, capacity 512) at ./tiny_dnn/network.h:1057
#6  0x00000000004a965f in tiny_dnn::network<tiny_dnn::graph>::fit<tiny_dnn::cross_entropy, tiny_dnn::adam, AutoScala_Deep::DoTraining()::<lambda()>, AutoScala_Deep::DoTraining()::<lambda()>, std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > >, unsigned int>(tiny_dnn::adam &, const std::vector<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > >, std::allocator<std::vector<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> >, std::allocator<std::vector<float, tiny_dnn::aligned_allocator<float, 64ul> > > > > > &, const std::vector<unsigned int, std::allocator<unsigned int> > &, size_t, int, AutoScala_Deep::<lambda()>, AutoScala_Deep::<lambda()>, bool, int, const std::vector<unsigned int, std::allocator<unsigned int> > &) (this=0x7fffffffdce0, 
    optimizer=..., inputs=std::vector of length 512, capacity 512 = {...}, desired_outputs=std::vector of length 512, capacity 512 = {...}, 
    batch_size=1, epoch=1, on_batch_enumerate=..., on_epoch_enumerate=..., reset_weights=false, n_threads=8, 
    t_cost=std::vector of length 0, capacity 0) at ./tiny_dnn/network.h:330
#7  0x00000000004a94ec in AutoScala_Deep::DoTraining (this=0x7fffffffdbd0) at ASD_train_arch.cpp:134
#8  0x0000000000405eb2 in main (argc=2, argv=0x7fffffffded8) at ASD_Main.cpp:23

![graph](https://cloud.githubusercontent.com/assets/6822038/25718742/3ae3d14e-30d5-11e7-9e15-436b45bc6a7d.png)

**Comment by shivak7** [comment]
So I have traced the problem down to: 
**normalize_tensor(desired_outputs, output_tensor)** which calls:
**net_.label2vec(inputs, vec)** which calls:
**label2vec(&labels[0], static_cast<serial_size_t>(labels.size()), vec)** @ nodes.h L194
which calls:
**serial_size_t outdim = out_data_size()** @nodes.h L181, the debugger prints outdim = 32767!!!
Finally **out_data_size()** @ layer.h L133 calls sumif. 

The segfault seems to happen at this point. Trying to print in_shape() or out_shape() causes a segfault within gdb!!!
Not sure why... Would appreciate any help at this point @beru.

**Comment by beru** [comment]
@shivak7 Can you post minimum reproducable code & data so that I can debug.

**Comment by shivak7** [comment]
@beru Sure. I'll try to get you a minimal code and a data set. 
In the meantime I wanted to ask how would one typically train/test a graph network with multiple input branches? I noticed that only the fit function even allowed tensor_t inputs. Train and the test functions do not.

**Comment by shivak7** [comment]
@beru I have condensed the code into a single file and am attaching it here. 
Unfortunately my dataset (65MB) is organized in a way that makes it difficult to reduce, so I am providing a link to the data files that will expire in 2 days:https://expirebox.com/download/ee50b4a122967dc6f91071c5dfda7dcb.html

Note that all the data folders should be in the same directory as example.cc 

Hope this helps. Thanks!
[example.zip](https://github.com/tiny-dnn/tiny-dnn/files/978042/example.zip)

**Comment by beru** [comment]
@shivak7 The layer instances created on stack memory in `Arch2Branched` can't be accessed later so I changed them to global variables.
[example.cc.zip](https://github.com/tiny-dnn/tiny-dnn/files/980167/example.cc.zip)

**Comment by shivak7** [comment]
@beru Thanks! I'll try debugging it a bit more... Did you find anything else that may indicate the segfault?

**Comment by shivak7** [comment]
I just realized what you meant in your previous comment. Maybe if a pseudo root-node is created above the first input layers in a tree structure (for graphs), it should still be possible for each layer to have arrays of generic base class layer pointers to handle branching in-out. Would the layer instances be accessible then if defined and linked together locally in a function?

**Comment by beru** [comment]
I'm not sure of clever and simpler ways to let the library manage ownerships of graph network nodes.
